src/network/socket_wrapper.hpp

```diff
@@ -36,6 +35,13 @@
 #include <sys/types.h>
 #include <unistd.h>
 
+// ifaddrs.h is not available on Solaris 10
+#if defined(sun) || defined(__sun)
```

Comment:
Looks like two checks (here and in `ifaddrs_patch.h` file) are excess.
Also, please see comment above about more complicated `if` condition for Solaris.

Reply:
I think you might be mistaken. I have not added this check in `ifaddrs_patch.h`

![image](https://user-images.githubusercontent.com/7608904/95288692-91a59d80-082e-11eb-8b41-9bee66db953f.png)

I do think it is necessary in both `ifaddrs_patch.cpp` and `socket_wrapper.hpp`. The check in `ifaddrs_patch.cpp` prevents compiling that patch on non-Solaris systems. The check here in `socket_wrapper.hpp` ensures that the correct header is included for `ifaddrs` based on whether or not you are on Solaris

Reply:
I just added an error to an example wrapped in `\donttest{}`, then re-generated the `{pkgdown}` site locally and I can see that it was run!

![image](https://user-images.githubusercontent.com/7608904/95289648-eba76280-0830-11eb-8198-4907bcfb6ed4.png)

Reply:
Yeah, I confused with `ifaddrs_patch.cpp`. If it was `*.hpp`, it will be able to do something like
```cpp
// ifaddrs_patch.hpp
#if (defined(sun) || defined(__sun)) && (defined(__SVR4) || defined(__svr4__))
  ...
  <actual patch>
  ...
#else
  #include <ifaddrs.h>
#endif
```
```cpp
// socket_wrapper.hpp
#include "ifaddrs_patch.hpp"
```
I believe.

Reply:
I thought so too, but when I first tried that it didn't quite work. Moving to its own `.cpp` file, which produces its own object, fixed it. https://github.com/microsoft/LightGBM/pull/3338#issuecomment-698119189

Reply:
Ah, I see! Thanks for the pointer! This discussion is getting bigger and bigger, can't remember all of discussed points, sorry 😞 .

> And there is an extern keyword. I remember that the implementation of extern var/function should be in c/cpp file.

Anyway, it works and it is fine! Not a big deal to have two files and two `ifdef`s, I believe 🙂 .